### DISCUSSION TITLE
Need both BLE and RF in ESP32 firmware

### DISCUSSION THREAD

**Discussion by mailminhthang** [arendst/Tasmota#15943] [Q&A] [answered]
Hi, I have an ESP32 board.
I can use bluetooth (BLE) functions by using firmware tasmota32-bluetooth.bin (but does not support RF)
I can use RF receiver functions by using firmware tasmota32.bin, set pin to RFRECV (106) (but does not support BLE)

If I want both functions: BLE and RF receiver, so how can I customize firmware to support both ?

Thanks

**Comment by barbudor** [comment, marked as answer]
You need to:
- add all needed BLE and RF requirements (#define USE_xxxx) in `user_config_override.h`
- add a custom environment in `platformio_tasmota_cenv.ini` (if you don't already have that file, copy the sample  `platformio_tasmota_cenv_sample.ini` to `platformio_tasmota_cenv.ini` 
```
[env:tasmota32-blerf]
extends                 = env:tasmota32_base
build_flags             = ${env:tasmota32_base.build_flags}
lib_extra_dirs          = lib/libesp32, lib/libesp32_div, lib/lib_basic, lib/lib_i2c, lib/lib_ssl
```
- build `tasmota32-blerf`

That should be ok

**Reply by pkkrusty** [reply]
Haha tasmota32-blerf. It's a funny sound in English, not sure about German.

**Reply by mailminhthang** [reply]
> You need to:
> 
> * add all needed BLE and RF requirements (#define USE_xxxx) in `user_config_override.h`
> * add a custom environment in `platformio_tasmota_cenv.ini` (if you don't already have that file, copy the sample  `platformio_tasmota_cenv_sample.ini` to `platformio_tasmota_cenv.ini`
> 
> ```
> [env:tasmota32-blerf]
> extends                 = env:tasmota32_base
> build_flags             = ${env:tasmota32_base.build_flags}
> lib_extra_dirs          = lib/libesp32, lib/libesp32_div, lib/lib_basic, lib/lib_i2c, lib/lib_ssl
> ```
> 
> * build `tasmota32-blerf`
> 
> That should be ok

Hi Bardudor
It's true that I am newbie for modify firmware, I just only download firmware (BIN file) and flash to ESP32 and then config via Tasmota web interface.
Is there any already customized firmware like that in anywhere and I can download it ?

Thanks

**Reply by barbudor** [reply]
no, you have to build your own
and this specific case will require VSCode+PlatformIO
I'm not sure you can do it with TasmoCompiler

Sincerely, you'd better to separate the 2 system using 2 boards. That will save you a lot of trouble.
